### PR TITLE
Convert SmartScalar to TypeScript

### PR DESCRIPTION
**Pull request by shaunlebron**
Step 1 of https://github.com/metabase/metabase/issues/33411

> [!Important]
> To review this PR:
> ⚙️ Hide whitespace in the diff.
> 👀 Look at the diffs of the individual commits.

One commit is for converting the component from `class` to `function`.

The other commit is for adding and fixing type annotations.

### CODE REVIEW

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.tsx
Comment (shaunlebron) [inline frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.tsx:300]: TypeScript couldn’t recognize this `responsive` property, nor could I.  I looked around and found nothing else using or referencing it.

File: frontend/src/metabase/visualizations/lib/settings/column.js
Comment (shaunlebron) [inline frontend/src/metabase/visualizations/lib/settings/column.js:29]: TS somehow infers `getColumns: never[]` unless I destructure the args outside the function signature.  I’m not sure why.

Comment (shaunlebron) [reply frontend/src/metabase/visualizations/lib/settings/column.js:29]: It looks like TS had trouble parsing the type of `[].concat(...series.map(...))` but it worked when I used `series.flatMap`.  I’ll make this change

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.tsx
Comment (EmmadUsmani) [inline frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.tsx:154]: I think we messed up when typing `settings.column`'s return value, it seems like we defined it as something completely different than it actually is

```typescript
export type ComputedVisualizationSettings = VisualizationSettings & {
  column?: (
    col: RemappingHydratedDatasetColumn,
  ) => RemappingHydratedDatasetColumn;
};
```

@shaunlebron I think it's fine to keep your typecast here since you seen to have the right type (although it seems `OptionsType` is missing a few properties like `currency` that are present on `formatOptions`, so it might be worth adding those.

@alxnddr Let's try to fix the return type of the `settings.column` method.

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.tsx
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.tsx:300]: I couldn't find anything either, seems like it's safe to remove. @kamilmielnik do you remember why you added the `responsive` prop in https://github.com/metabase/metabase/pull/31722?

File: frontend/src/metabase/visualizations/types/visualization.ts
Comment (EmmadUsmani) [inline frontend/src/metabase/visualizations/types/visualization.ts:118]: Similar to with `staticProps`, let's name this something like `VisualizationProperties` instead to avoid confustion with React props.

File: frontend/src/metabase/visualizations/types/visualization.ts
Comment (EmmadUsmani) [inline frontend/src/metabase/visualizations/types/visualization.ts:156]: Do you plan on doing this in this PR or later?

**Review by EmmadUsmani — APPROVED** [review]
Overall looks good! I did have some nit comments for a few things, please take a look at them before merging.

File: frontend/src/metabase/visualizations/lib/settings/column.js
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/lib/settings/column.js:29]: Do you still plan on changing this?

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.tsx
Comment (alxnddr) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.tsx:154]: Oh yes, the return type is incorrect. It is also not `OptionsType` but `ColumnSettings` which is probably missing.
The `OptionsType` is a part of the formatting code interface

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.tsx:300]: It must have been a mistake: 07d086cb072aa9cb50ac7d8a4d238cadc0d2956f
Safe to remove :+1:

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx
Comment (shaunlebron) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx]: 2a8e4c1d27655dbcf0b96f9c4c15aae1503545fb

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx
Comment (shaunlebron) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx]: 66fbffd1a83e57f4fc3a7a18413c5343e2bb7a1a

File: frontend/src/metabase/visualizations/types/visualization.ts
Comment (shaunlebron) [reply frontend/src/metabase/visualizations/types/visualization.ts]: 66fbffd1a83e57f4fc3a7a18413c5343e2bb7a1a

File: frontend/src/metabase/visualizations/types/visualization.ts
Comment (shaunlebron) [reply frontend/src/metabase/visualizations/types/visualization.ts:156]: This is sasha’s comment that I just moved down.  I’m not sure what it means.

Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/types/visualization.ts:156]: Ah got it. It looks like you ended up doing this though since you made the `Visualization` type based on `React.ComponentType<VisualizationProps>`. I think we can remove this comment now, cc @alxnddr for confirmation.

Comment (alxnddr) [reply frontend/src/metabase/visualizations/types/visualization.ts:156]: @shaunlebron I wonder why you did not add `name: string;` property to the `Visualization` type which looks like the shortest path, wdyt about it? The `VisualizationProperties` looks a bit ambiguous in relation to VisualizationProps.

The idea behind my TODO comment was to separate visualization definitions from the component: replace `Object.assign(component, { checkRenderable: () => {}, ... })` with `{ component: PieChart, checkRenderable: () => {}, ... }`

**Comment by shaunlebron** [comment]
CI is blocked by type errors I’ve already resolved in a giant PR that I need to split up.  I’ll close this and reopen in a smaller PR.

**Pull request closed without merging**
